SetBackgroundColour overridden on wxAuiToolBar

ThueAndersen opened the pull request:
When the toolbar is set to draw a plain background instead of the
default gradient background (wxAUI_TB_PLAIN_BACKGROUND) there is no way
to control the colour.
Calling SetBackgroundColour on the toolbar currently sets the colour on
the wxAuiToolBar, but the art provider always uses the default system
color.
Setbackgroundcolour now forwards the colour to the art provider.

Getter functions not modifying the object is now const.

ThueAndersen committed:
SetBackgroundColour overridden on wxAuiToolBar

When the toolbar is set to draw a plain background instead of the
default gradient background (wxAUI_TB_PLAIN_BACKGROUND) there is no way
to control the colour.
Calling SetBackgroundColour on the toolbar currently sets the colour on
the wxAuiToolBar, but the art provider always uses the default system
color.
Setbackgroundcolour now forwards the colour to the art provider.

Getter functions not modifying the object is now const.

vadz commented:
I'm afraid we can't add `const` to these methods as they can be overridden in the existing code and so doing this would silently break it. Forwarding the colour to the art provider looks fine to me though, so if you could please resubmit with just this part of the changes, I'd gladly apply it. TIA!

ThueAndersen committed:
Revert "SetBackgroundColour overridden on wxAuiToolBar"

This reverts commit bfcfadebb18f60edffb31ef1a8f7a54c8e29a28d.

ThueAndersen committed:
SetBackgroundColour overridden on wxAuiToolBar

When the toolbar is set to draw a plain background instead of the
default gradient background (wxAUI_TB_PLAIN_BACKGROUND) there is no way
to control the colour.
Calling SetBackgroundColour on the toolbar currently sets the colour on
the wxAuiToolBar, but the art provider always uses the default system
color.
Setbackgroundcolour now forwards the colour to the art provider.

allanjos commented:
[wxWidgets-auibar-20151126_2342.diff.txt](https://github.com/wxWidgets/wxWidgets/files/45405/wxWidgets-auibar-20151126_2342.diff.txt)

Follow a modification removing hard-coded color and passing toolbar style to art provider to be possible to decide how to draw based on it.

To set toolbar background you can call something like this:

    class ToolbarArtProvider : public wxAuiDefaultToolBarArt
    {
    public:
       ToolbarArtProvider() {
            SetBaseColour(wxSystemSettings::GetColour(wxSYS_COLOUR_WINDOW));
        }
    };

An set it to toolbar:

    ToolbarArtProvider *artProvider = new ToolbarArtProvider();

    toolBar->SetArtProvider(artProvider);


Thank you.

vadz commented:
@allanjos Sorry, I'm confused, is this patch proposed as an addition to or replacement for this PR?

allanjos commented:
As an addition, if it is correct. It is what I'm using to fix the issue with page themed background inside notebooks, I set background to wxSYS_COLOUR_WINDOW to match the background colors of the toolbar with the panel / scrolled window behind it.

Thank you.

vadz commented:
Sorry, this still doesn't answer my question and I still don't know whether I should the patch in this PR and your patch or do you mean your patch to _replace_ the patch here? FWIW the (updated version of the) patch in this PR looks fine to me, is your patch still needed even if it's applied? Alternatively, do you think the patch in the PR shouldn't be applied and your patch should and, if so, why?

allanjos commented:
My modification was not written based on the patch of that pull request. It implements something similar but add some code to maintain the compatibility with the current implementation, like the toolbar style being passed to the art provider to not alter the color on GetInitialBaseColor() for plain background. It is only a git diff from my modified code from master that I'm using on some applications for Linux and Windows. I've posted this to the author reuse something from it, if he wants to.

BR.

vadz commented:
I still don't know what to do with this, so I'm afraid I have to close it. If either @ThueAndersen or @allanjos can explain what exactly is proposed here and what problem exactly does it fix, please reopen it. Thanks!
